### PR TITLE
Share buffer for liveness::simulate_block

### LINKED ISSUE #51818: [nll] share buffer for `liveness::simulate_block`
The `simulate_block` helper takes the set of live variables at the exit of the block and computes the live variables at each point in the block:

https://github.com/rust-lang/rust/blob/764232cb2a8407c72b9fea68835e686240e30ef3/src/librustc_mir/util/liveness.rs#L161-L167

It works by creating a temporary bitset that contains the final value from the end of the block to start:

https://github.com/rust-lang/rust/blob/764232cb2a8407c72b9fea68835e686240e30ef3/src/librustc_mir/util/liveness.rs#L171-L172

**Rather than re-allocating this buffer, it would be nice to share it across invocations of `simulate_block`.** This makes sense because `simulate_block` is really only invoked from one place (apart from debugging output), which is [the method `add_liveness_constraint`](https://github.com/rust-lang/rust/blob/master/src/librustc_mir/borrow_check/nll/type_check/liveness.rs#L82). That method is invoked from a simple loop over all basic blocks:

https://github.com/rust-lang/rust/blob/master/src/librustc_mir/borrow_check/nll/type_check/liveness.rs#L52-L54

We could therefore have `simulate_block` take some extra "buffer" parameter (which would be stored on the stack and passed to `add_liveness_constraint`).

According to recent profiles, this fn accounts for about 2% of MIR borrowck, though, so this may not be worth it. Still, we're at the point where "every little bit helps!"

### PR DESCRIPTION
**Pull request by spastorino**
Closes #51818

r? @nikomatsakis

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
Perf queued but it'll be .. a while.

**Comment by nikomatsakis** [comment]
Perf URL: http://perf.rust-lang.org/compare.html?start=860d169474acabdc53b9a698f8ce02eba7e0daeb&end=00bcc44fb92c28465c727881355c3235a56a4045&stat=instructions%3Au

Results still not available :'(

**Comment by nikomatsakis** [comment]
r=me presuming perf results are good (will need a rebase though)

File: src/librustc_mir/util/liveness.rs [hunk 1]
```diff
@@ -110,19 +110,26 @@ impl LivenessResults {
             ),
         }
     }
+
+    pub fn make_simulate_buffer<'tcx>(&self, mir: &Mir<'tcx>) -> LocalSet {
```
Comment (nikomatsakis) [inline src/librustc_mir/util/liveness.rs]: can you add a comment here about how this buffer is held across invokes of `simulate_block` for performance reasons?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

**Comment by nikomatsakis** [comment]
Looking at [the perf results](https://perf.rust-lang.org/compare.html?start=860d169474acabdc53b9a698f8ce02eba7e0daeb&end=00bcc44fb92c28465c727881355c3235a56a4045&stat=instructions%3Au), @spastorino and I decided that this PR isn't really helping, and it seems to make the code mildly messier, so we're gonna close it.

**Pull request closed without merging**